presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java

```diff
@@ -460,8 +471,13 @@ public void start()
                         Thread.currentThread(),
                         timeoutThreadExecutor,
                         getQueryAnalyzerTimeout(getSession()))) {
-                    // create logical plan for the query
-                    plan = createLogicalPlanAndOptimize();
+                    // If planFuture has not started, cancel and build plan in current thread
+                    if (planFuture != null && !planFutureLocked.compareAndSet(false, true)) {
```

Comment:
From the discussion at https://github.com/prestodb/presto/pull/23649#discussion_r1765583206, trying to call `planFuture.cancel(false)` will mark the future as cancelled, but the thread will continue to run. Using the AtomicBoolean `planFutureLocked` will effectively cancel the planFuture if set here first, then the future result (null) is discarded.